### PR TITLE
New migration system in v3

### LINKED ISSUE #1189: [v3] Migrations backwards compatibility
According to [our own doc](https://procrastinate.readthedocs.io/en/stable/howto/production/migrations.html), migrations named with version X are the migrations you can apply once you run on version X.
According to our doc, people on v2.13 wishing to upgrade without disruption would:
- Change the code to v3.0
- Apply the v3.0 migrations
- Change the code to v3.1
- Apply the v3.1 migrations
All the while their instance is running and processing jobs.

Compatibility matrix:

| Code ➡️<br>⬇️ Schema | v2.13 | v3.0 | v3.1 |
|--------|--------|--------|--------|
| v2.13 | ✅ | ✅ | ❌ |
| v3.0 | ❌ | ✅ | ✅ |
| v3.1 | ❌ | ❌ | ✅ |

❌ means we don't have to make it compatible, but if it happens to be, it's not a problem, of course

We need to decide what we put in v3.0 and v3.1

Note: the change of `status = aborting -> abort_requested` might need a couple of trigger just for the time of the dual compatibility (adding the `abort_requested column and update its value based on the status, then add a trigger that synchronizes the 2 columns, and in the next release, delete the triggers and the column)

### PR DESCRIPTION
**Pull request by medihack**
Closes #1189

Introduce a new migration process: pre-migrations and post-migrations.
- They will be tied to the corresponding release
- Reworked v3 migrations with this new template in mind
- They allow us to provide a much saner path for blue-green deployment
- The process is currently not tested (but nor was the previous system)
- We will need to be careful when we do a release to rename the migrations in ongoing PRs (as today)
- We (hopefully) won't have to do multiple releases just for the sake of blue-green compatibility.
- I'm adding versioning in the name of all internal objects (functions, triggers, etc) (as mentionned [here](https://github.com/procrastinate-org/procrastinate/issues/1189#issuecomment-2376450912))

So the idea is that:
- Procrastinate v2 with the pre-migrations & without the post-migrations should work
- Procrastinate v3 with the pre-migrations & without the post-migrations should work
- Procrastinate v3 with all migrations should be what we currently have in v3

(I'm 80% sure that I have missed a few corner cases)

### Successful PR Checklist:

- [ ] Tests
  - [ ] (not applicable?)
- [x] Documentation
  - [ ] (not applicable?)

#### PR label(s): 
  - [x] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20breaking%20%F0%9F%92%A5
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20feature%20%E2%AD%90%EF%B8%8F
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20bugfix%20%F0%9F%95%B5%EF%B8%8F
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20miscellaneous%20%F0%9F%91%BE
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20dependencies%20%F0%9F%A4%96
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20documentation%20%F0%9F%93%9A

### CODE REVIEW

**Comment by medihack** [comment]
@ewjoachim, in our session, I think we made a wrong assumption regarding function renaming. Function bodies are stored as strings, so when a function is renamed and used by another function, the function using the renamed function must be recreated, too. This is not true for triggers (or tables in case of used types) as those store the references.
So I did it in very small steps to make it comprehensible, and everything is renamed now.

File: procrastinate/sql/queries.sql [hunk 1]
```diff
@@ -5,17 +5,17 @@
...
 -- Create a periodic job if it doesn't already exist, and delete periodic metadata
 -- for previous jobs in the same task.
-SELECT procrastinate_defer_periodic_job(%(queue)s, %(lock)s, %(queueing_lock)s, %(task_name)s, %(periodic_id)s, %(defer_timestamp)s, %(args)s) AS id;
+SELECT procrastinate_defer_periodic_job_v1(%(queue)s, %(lock)s, %(queueing_lock)s, %(task_name)s, %(priority)s, %(periodic_id)s, %(defer_timestamp)s, %(args)s) AS id;
 
 -- fetch_job --
 -- Get the first awaiting job
 SELECT id, status, task_name, priority, lock, queueing_lock, args, scheduled_at, queue_name, attempts
-    FROM procrastinate_fetch_job(%(queues)s);
+    FROM procrastinate_fetch_job_v1(%(queues)s::varchar[]);
```
Comment (medihack) [inline procrastinate/sql/queries.sql:18]: I adopted this from your code ... why have you explicitly added the `::varchar[]` here?

Comment (medihack) [reply procrastinate/sql/queries.sql:18]: Ok, thanks. I just leave it then that way.

**Review by ewjoachim — APPROVED** [review]
Awesome work :) I'm still trying to find some time to test this with v2 & v3 with only pre. Life is a bit hectic at the moment so I can't commit on a date. Still, I'm very appreciative of your work :)

**Comment by medihack** [comment]
> Awesome work :) I'm still trying to find some time to test this with v2 & v3 with only pre. Life is a bit hectic at the moment so I can't commit on a date. Still, I'm very appreciative of your work :)

Just take your time. Let me know if there is anything specific I can help with. I will also look into how to test the pre schema with v2 and v3.

**Comment by medihack** [comment]
@ewjoachim  Unfortunately, I have difficulty checking if the pre-migration runs with even v3. For testing purposes, I deleted the post-migration, ran all migrations on an empty database, and dumped the resulting schema with `pg_dump`. Then I replaced our `schema.sql` with that dump and ran the integration and acceptance tests, but I ended up with many (for me cryptic) errors. Was my approach correct? Have you had something different in mind? Do you perhaps have time for a video conference next week or so to solve this together?

**Comment by ewjoachim** [comment]
I'd love that :) let's schedule this together !
